addons/account/models/account_move.py

```diff
@@ -3715,14 +3715,14 @@ def _check_constrains_account_id_journal_id(self):
 
     @api.constrains('account_id', 'tax_ids', 'tax_line_id', 'reconciled')
     def _check_off_balance(self):
-        for line in self:
-            if line.account_id.internal_group == 'off_balance':
-                if any(a.internal_group != line.account_id.internal_group for a in line.move_id.line_ids.account_id):
-                    raise UserError(_('If you want to use "Off-Balance Sheet" accounts, all the accounts of the journal entry must be of this type'))
-                if line.tax_ids or line.tax_line_id:
-                    raise UserError(_('You cannot use taxes on lines with an Off-Balance account'))
-                if line.reconciled:
-                    raise UserError(_('Lines from "Off-Balance Sheet" accounts cannot be reconciled'))
+        off_balance_line = self.search([('id', 'in', self.ids), ('account_id.internal_group', '=', 'off_balance'), ], limit=1)
+        non_off_balance_line = self.search([('id', 'in', self.ids), ('account_id.internal_group', '!=', 'off_balance'), ], limit=1)
```

Comment:
@hbto I don't think this is the same behaviour as before.  And 'in self.ids' can not be very efficient as well on large recordsets than the cache.  The O(n2) could probably be avoided by a search however to see if there are other lines in the same account move...  But then you need a lot of off-balance lines already.   @william-andre

Reply:
@jco-odoo Thanks for your feedback I am due to profile before and after this change.

Regards.

Reply:
The search is most likely way worse than iterating on `self` on a big database. Unless you can provide more information about the performance than what you did, it is a no go.
By the way, the condition of the second and third are `O(n)` too.

Also, like @jco-odoo said, this is not the same behavior as before because you are not taking into account the grouping by `account.move`

What could be done is aggregate all the account's internal groups by move in a dict mapping moves to sets, in one pass. Then checking if each set contains `off_balance`, then its length must be 1.
* One independent loop in `O(#lines)`
* One independent loop in `O(#moves)`

It will still use the cache when possible (if all the lines of all the moves are already cached) too.

But I really feel like it is over-engineered if it is not causing real trouble.

Reply:
I have overlooked the grouping by move. To be taken into account in the next iteration